doc/developer/design/20220314_linear_or_eventual.md

```diff
@@ -0,0 +1,115 @@
+
+## Description
+
+### Background
+
+Sources currently produce messages that contain offsets. Those offsets are written down, associated with a timestamp, as a _timestamp binding_.
+When you do a peek (read) against some source. A recent timestamp is chosen, the index is consulted for data that is valid in respect to that timestamp
+and the _timestamp bindings_.
+
+### Linearizable reads
+In the case of `LINEARIZABLE` read (see below for how to express this in sql), peeks will be augemented to work as follows
+
+1. Before issuing the peek, we will first ask each source involved in the peek: "what is your max offset"
+2. We will wait until a max offset is available at some timestamp `t` (i.e., written down as a timestamp binding)
+3. And then issue the peek for timestamp `t`.
+
+In the case of "unmaterialized" sources, the temporary index will be created at the beginning of this process and not removed until the end.
+
+### Eventually consistent reads
+This is the default behavior of peeks, and will be augmented as follows:
+
+1. Before issuing the peek, we will look for the latest available timestamp for all involved sources
+2. And then issue the peek at that timestamp
+
+This is the standard `since` and `uppser` logic that already exists in `materialized`. Additionally, "unmaterialized"
+sources won't have access to do this. A query involving this
+
+In the case of "unmaterialized" sources, there are 2 important notes:
+- the temporary index will be created at the beginning of this process and not removed until the end.
+- When an unmaterialized source is created, a background special _index_ will be created, which will progress
+the source and produce only offsets that are bound to timestamps. This allows EC reads against unmaterialized sources
+to consistently obtain new data.
+
+### Platform considerations
+
+In the world of Platform, when performing a _linearizable_ read, we need to be specific about what we mean by
+steps 1&2. The way these will operate is:
+
+1. `STORAGE` will offer an api that simply takes in the list of involved sources, and blocks until it returns a timestamp `t`
+2. `STORAGE`, upon receiving this request, will ask the sources for their max offets, and then wait for a timestamp binding
+that contains all those offsets.
+3. And then issue the peek for timestamp `t`.
+
+### Extensions
+
+This document currently describes a situation where querying unmaterialized sources requires waiting for _linearizability_.
+A natural extension of this would be to add a _heartbeating_ mechanism that continuously records offsets, but not data. This
+would be a special version of a long-living index.
+
+In the Platform world, this point may be unimportant, as sources may always be making incremental progress, regardless of their
+materialization status. (TODO(guswynn): confirm this with petros)
+
+
+## Proposed syntax changes
+
+`SELECT` statements will be altered to add an optional `LINEARIZABLE` after the `SELECT keyword.
```

Comment:
Spitballing, but might it make sense closer to `AS OF` or is that a misuse/conflation?

Reply:
Hmm, probably a good pint
`AS OF` is an extension at the end of the the outermost statement, we should probably add something there with the same semantics (https://github.com/MaterializeInc/materialize/blob/main/src/sql-parser/src/parser.rs#L272), probably just an optional `LINEARIZABLE`

Reply:
ugh, I think github lost my comments, but yeah, I agree that this makes more sense!